File: src/strategy_ai_bridge.py
#!/usr/bin/env python3

import rclpy
import numpy as np
from scipy.spatial.transform import Rotation as R
from rclpy.node import Node

from std_msgs.msg import Int64MultiArray
from geometry_msgs.msg import Pose2D, PoseArray, PoseStamped

from override_sim.msg import (
    FieldElementArray, GoalArray, ToggleArray, WorldState,
)


class StrategyAIBridge(Node):
    """Subscribes to every simulation topic and packs one override_sim/WorldState
    message for the Strategy AI node on /sai_input."""

    def __init__(self):
        super().__init__('sai_bridge')

        self.create_subscription(
            Int64MultiArray, '/game_score', self.score_cb, 10)
        self.create_subscription(
            GoalArray, '/goals', self.goals_cb, 10)
        self.create_subscription(
            ToggleArray, '/toggles', self.toggles_cb, 10)
        self.create_subscription(
            PoseArray, '/otto_pose', self.robot_pose_cb, 10)
        self.create_subscription(
            Int64MultiArray, '/robot_elements', self.intake_cb, 10)
        self.create_subscription(
            FieldElementArray, '/field_objects', self.field_elements_cb, 10)
        self.create_subscription(
            Int64MultiArray, '/loaders', self.loader_cb, 10)
        self.create_subscription(
            Int64MultiArray, '/elements_remaining', self.elements_left_cb, 10)
        self.create_subscription(
            PoseStamped, '/opponent/pose', self.opponent_pose_cb, 10)

        self.create_timer(1.0, self.update_sai_world_state)
        self.to_sai = self.create_publisher(WorldState, '/sai_input', 10)

        self.world_state = WorldState()

    @staticmethod
    def pose2d_from_pose_array(msg: PoseArray):
        pose = Pose2D()
        pose.x = float(msg.poses[-1].position.x)
        pose.y = float(msg.poses[-1].position.y)
        q = msg.poses[-1].orientation
        quat = np.array([q.x, q.y, q.z, q.w])
        quat = quat / (np.linalg.norm(quat) or 1.0)
        pose.theta = float(R.from_quat(quat).as_euler('xyz')[2])
        return pose

    def score_cb(self, msg):
        self.world_state.score = msg

    def goals_cb(self, msg):
        self.world_state.goals = msg

    def toggles_cb(self, msg):
        self.world_state.toggles = msg

    def robot_pose_cb(self, msg):
        self.world_state.robot_pose = self.pose2d_from_pose_array(msg)

    def intake_cb(self, msg):
        self.world_state.robot_intake = msg

    def field_elements_cb(self, msg):
        self.world_state.field_elements = msg

    def loader_cb(self, msg):
        self.world_state.loaders = msg

    def elements_left_cb(self, msg):
        self.world_state.elements_left = msg

    def opponent_pose_cb(self, msg: PoseStamped):
        pose = Pose2D()
        pose.x = msg.pose.position.x
        pose.y = msg.pose.position.y
        q = msg.pose.orientation
        quat = np.array([q.x, q.y, q.z, q.w])
        quat = quat / (np.linalg.norm(quat) or 1.0)
        pose.theta = float(R.from_quat(quat).as_euler('xyz')[2])
        self.world_state.opponent_pose = pose

    def update_sai_world_state(self):
        self.world_state.header.stamp = self.get_clock().now().to_msg()
        self.world_state.header.frame_id = 'map'
        self.to_sai.publish(self.world_state)


def main(args=None):
    rclpy.init(args=args)
    node = StrategyAIBridge()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/strategy_ai_bridge.cpp
// strategy_ai_bridge.cpp - subscribes to every simulation topic and packs one
// override_sim/WorldState message for the Strategy AI node on /sai_input.
#include <chrono>
#include <memory>
#include <utility>

#include <geometry_msgs/msg/pose2_d.hpp>
#include <geometry_msgs/msg/pose_array.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int64_multi_array.hpp>

#include "override_sim/field_constants.h"
#include "override_sim/msg/field_element_array.hpp"
#include "override_sim/msg/goal_array.hpp"
#include "override_sim/msg/toggle_array.hpp"
#include "override_sim/msg/world_state.hpp"

namespace override_sim
{

class StrategyAIBridge : public rclcpp::Node
{
public:
  StrategyAIBridge() : Node("sai_bridge")
  {
    score_sub_ = this->create_subscription<std_msgs::msg::Int64MultiArray>(
      "/game_score", 10,
      [this](const std_msgs::msg::Int64MultiArray::SharedPtr msg) {
        world_state_.score = *msg;
      });
    goals_sub_ = this->create_subscription<GoalArray>(
      "/goals", 10,
      [this](const GoalArray::SharedPtr msg) { world_state_.goals = *msg; });
    toggles_sub_ = this->create_subscription<ToggleArray>(
      "/toggles", 10,
      [this](const ToggleArray::SharedPtr msg) { world_state_.toggles = *msg; });
    robot_pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseArray>(
      "/otto_pose", 10,
      [this](const geometry_msgs::msg::PoseArray::SharedPtr msg) {
        world_state_.robot_pose = pose2d_from_pose_array(msg);
      });
    intake_sub_ = this->create_subscription<std_msgs::msg::Int64MultiArray>(
      "/robot_elements", 10,
      [this](const std_msgs::msg::Int64MultiArray::SharedPtr msg) {
        world_state_.robot_intake = *msg;
      });
    field_sub_ = this->create_subscription<FieldElementArray>(
      "/field_objects", 10,
      [this](const FieldElementArray::SharedPtr msg) {
        world_state_.field_elements = *msg;
      });
    loaders_sub_ = this->create_subscription<std_msgs::msg::Int64MultiArray>(
      "/loaders", 10,
      [this](const std_msgs::msg::Int64MultiArray::SharedPtr msg) {
        world_state_.loaders = *msg;
      });
    elements_left_sub_ = this->create_subscription<std_msgs::msg::Int64MultiArray>(
      "/elements_remaining", 10,
      [this](const std_msgs::msg::Int64MultiArray::SharedPtr msg) {
        world_state_.elements_left = *msg;
      });
    opponent_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/opponent/pose", 10,
      [this](const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        world_state_.opponent_pose.x = msg->pose.position.x;
        world_state_.opponent_pose.y = msg->pose.position.y;
        world_state_.opponent_pose.theta = YawFromQuat(
          msg->pose.orientation.x, msg->pose.orientation.y,
          msg->pose.orientation.z, msg->pose.orientation.w);
      });

    timer_ = this->create_wall_timer(
      std::chrono::seconds(1), [this]() { update_sai_world_state(); });
    sai_pub_ = this->create_publisher<WorldState>("/sai_input", 10);
  }

private:
  static geometry_msgs::msg::Pose2D pose2d_from_pose_array(
    const geometry_msgs::msg::PoseArray::SharedPtr msg)
  {
    geometry_msgs::msg::Pose2D pose;
    if (msg->poses.empty()) {
      return pose;
    }
    const auto & p = msg->poses.back();
    pose.x = p.position.x;
    pose.y = p.position.y;
    pose.theta = YawFromQuat(
      p.orientation.x, p.orientation.y, p.orientation.z, p.orientation.w);
    return pose;
  }

  void update_sai_world_state()
  {
    world_state_.header.stamp = this->now();
    world_state_.header.frame_id = "map";
    sai_pub_->publish(world_state_);
  }

  WorldState world_state_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<WorldState>::SharedPtr sai_pub_;
  rclcpp::Subscription<std_msgs::msg::Int64MultiArray>::SharedPtr score_sub_;
  rclcpp::Subscription<GoalArray>::SharedPtr goals_sub_;
  rclcpp::Subscription<ToggleArray>::SharedPtr toggles_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseArray>::SharedPtr robot_pose_sub_;
  rclcpp::Subscription<std_msgs::msg::Int64MultiArray>::SharedPtr intake_sub_;
  rclcpp::Subscription<FieldElementArray>::SharedPtr field_sub_;
  rclcpp::Subscription<std_msgs::msg::Int64MultiArray>::SharedPtr loaders_sub_;
  rclcpp::Subscription<std_msgs::msg::Int64MultiArray>::SharedPtr
    elements_left_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr opponent_sub_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::StrategyAIBridge>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
